fix: node.js の punycode モジュールが使用されている場所がある問題

anatawa12 opened the pull request:
<!-- ℹ お読みください / README
PRありがとうございます！ PRを作成する前に、コントリビューションガイドをご確認ください:
Thank you for your PR! Before creating a PR, please check the contribution guide:
https://github.com/misskey-dev/misskey/blob/develop/CONTRIBUTING.md
-->

## What
<!-- このPRで何をしたのか？ どう変わるのか？ -->
<!-- What did you do with this PR? How will it change things? -->

Fixing #15246 but not closing since some dependency still uses

ついでに今後の事故をなくすため lint を追加しました。

## Why
<!-- なぜそうするのか？ どういう意図なのか？ 何が困っているのか？ -->
<!-- Why do you do it? What are your intentions? What is the problem? -->
Warningがうるさい。
多分モジュールがnodejsから削除されたら静になるけどまだそうならないので

## Additional info (optional)
<!-- テスト観点など -->
<!-- Test perspective, etc -->

CHANGELOGなくても良い機がする

## Checklist
- [ ] Read the [contribution guide](https://github.com/misskey-dev/misskey/blob/develop/CONTRIBUTING.md)
- [ ] Test working in a local environment
- [ ] (If needed) Add story of storybook
- [ ] (If needed) Update CHANGELOG.md
- [ ] (If possible) Add tests

anatawa12 committed:
fix: punycode.js が使用されていない場所がある問題

samunohito commented:
> SyntaxError: The requested module 'punycode/' does not provide an export named 'toASCII'

おお・・・？

anatawa12 commented:
`/`で終わってる場合、`main`を読む仕組みだったけどindex.jsを読むのが無効化されてるからだめとかかな。いい感じにパスにします。

anatawa12 commented:
importすべきは`punycode.es6.js`っぽいけど`punycode.es6.js`が`@types/punycode`にない...

anatawa12 committed:
fix: use punycode/punycode.js on backend

anatawa12 committed:
fix: use punycode/punycode.es6.js on backend

anatawa12 committed:
fix: d.ts missing declare keyword

anatawa12 commented:
commonjsの方だとエラーになったので、 'punycode/punycode.es6.js' のd.tsを生やして対処しました。

`@types/punycode`の方への修正後で投げておきます

anatawa12 commented:
https://github.com/mathiasbynens/punycode.js/pull/142 とりあえず .mjs を作る PR 投げたけど、 backend で punycode 使ってるの utility service だけなので、別のAPI叩くようにしたほうが良さそう。

anatawa12 committed:
chore: don't use punycode.js on backend

anatawa12 committed:
update pnpm-lock.yaml

anatawa12 commented:
backendではpunycode.jsを使わずnode:urlを使うことにしました

anatawa12 commented:
どうにかchecks通った(chromaticは原因なんだろ)

u1-liquid commented:
結局のところ、当方で使用している punycode パッケージを変更しても、どこかで参照されているため警告メッセージは消えません。
```
＜checksのログから＞
(node:3575) [DEP0040] DeprecationWarning: The `punycode` module is deprecated. Please use a userland alternative instead.
```
また、import 文で`/`を付けるトリックを使うよりも、`punycode.js`パッケージ（punycodeパッケージと同時に公開されています）を使用したほうがよろしいのではないかと思います。
https://github.com/MisskeyIO/misskey/pull/833/files#diff-8571ecd91584b00015b23695d3a6a164282636bb47bfbe46dca243bf9b4db773R61
https://github.com/MisskeyIO/misskey/pull/833/files#diff-8571ecd91584b00015b23695d3a6a164282636bb47bfbe46dca243bf9b4db773R109

anatawa12 committed:
chore: remove punycode.d.ts

anatawa12 committed:
chore: use punycode.js instead of punycode npm package

anatawa12 commented:
ありがとうございます。`punycode.js`を使用するようにしました。

warningが残る問題については依存関係の問題ですが、grepした限りだと tough-cookie@2.5.0 tr46@0.0.3 uri-js@4.4.1 psl@1.9.0 whatwg-url@5.0.0 が使用していたため、これらの関連パッケージの方を調査しますが、とりあえずmisskeyから使用をなくすことに価値はあると思います
`

```
@nestjs/core 10.4.7
└─┬ @nuxtjs/opencollective 0.3.2
  └─┬ node-fetch 2.7.0
    └─┬ whatwg-url 5.0.0
      └── tr46 0.0.3
@nestjs/testing 10.4.7
└─┬ @nestjs/core 10.4.7 peer
  └─┬ @nuxtjs/opencollective 0.3.2
    └─┬ node-fetch 2.7.0
      └─┬ whatwg-url 5.0.0
        └── tr46 0.0.3
@simplewebauthn/server 10.0.1
└─┬ cross-fetch 4.0.0
  └─┬ node-fetch 2.7.0
    └─┬ whatwg-url 5.0.0
      └── tr46 0.0.3
jsdom 24.1.1
└─┬ tough-cookie 4.1.4
  └── psl 1.9.0
nsfwjs 2.4.2
├─┬ @nsfw-filter/gif-frames 1.0.2
│ └─┬ get-pixels-frame-info-update 3.3.2
│   └─┬ request 2.88.2
│     ├─┬ har-validator 5.1.5
│     │ └─┬ ajv 6.12.6
│     │   └── uri-js 4.4.1
│     └─┬ tough-cookie 2.5.0
│       └── psl 1.9.0
└─┬ @tensorflow/tfjs 4.4.0 peer
  ├─┬ @tensorflow/tfjs-backend-cpu 4.4.0
  │ └─┬ @tensorflow/tfjs-core 4.4.0 peer
  │   └─┬ node-fetch 2.6.13
  │     └─┬ whatwg-url 5.0.0
  │       └── tr46 0.0.3
  ├─┬ @tensorflow/tfjs-backend-webgl 4.4.0
  │ ├─┬ @tensorflow/tfjs-backend-cpu 4.4.0
  │ │ └─┬ @tensorflow/tfjs-core 4.4.0 peer
  │ │   └─┬ node-fetch 2.6.13
  │ │     └─┬ whatwg-url 5.0.0
  │ │       └── tr46 0.0.3
  │ └─┬ @tensorflow/tfjs-core 4.4.0 peer
  │   └─┬ node-fetch 2.6.13
  │     └─┬ whatwg-url 5.0.0
  │       └── tr46 0.0.3
  └─┬ @tensorflow/tfjs-converter 4.4.0
    └─┬ @tensorflow/tfjs-core 4.4.0 peer
      └─┬ node-fetch 2.6.13
        └─┬ whatwg-url 5.0.0
          └── tr46 0.0.3
```

anatawa12 committed:
chore: bump psl to 1.15.0

anatawa12 committed:
chore: bump nsfwjs to 4.2.0

4.2.1 is not usable because of https://github.com/infinitered/nsfwjs/issues/904

anatawa12 committed:
chore: prevent loading node-fetch from tensorflow

anatawa12 committed:
chore: DOMWindow['document'] => Document

IDK why DOMWindow['document'] fails, but might be related to tsc internal complexity limit

anatawa12 committed:
fix: disable --trace-deprecation

syuilo committed:
Merge branch 'develop' into fix-punycode
